src/model/sprite.ts

```diff
@@ -0,0 +1,205 @@
+const createFlippedImg = (
+  inputCanvas: HTMLCanvasElement
+): HTMLCanvasElement => {
+  const [canvas, ctx, width] = G_model_createCanvas(
+    inputCanvas.width,
+    inputCanvas.height
+  );
+  ctx.translate(width, 0);
+  ctx.scale(-1, 1);
+  ctx.drawImage(inputCanvas, 0, 0);
+  return canvas;
+};
+
+// given a Sprite, create and return a new image from the sprite
+const spriteToCanvas = (sprite: Sprite): HTMLCanvasElement => {
+  const [, , , spriteWidth, spriteHeight] = sprite;
+  const [canvas, ctx] = G_model_createCanvas(spriteWidth, spriteHeight);
+  G_view_drawSprite(sprite, 0, 0, 1, ctx);
+  return canvas;
+};
+
+// load a set of sprites from an image, each sprite loaded with also have a set of rotated
+// and flipped variants
+const loadSpritesFromImage = (
+  spriteMap: SpriteCollection, // collection in which to put created sprites
+  image: HTMLImageElement | HTMLCanvasElement, // parent image
+  spritePrefix: string, // created sprites are named <spritePrefix>_<index>
+  spriteWidth: number,
+  spriteHeight: number
+) => {
+  const addSprite = (
+    name: string,
+    image: HTMLImageElement | HTMLCanvasElement,
+    x: number,
+    y: number,
+    w: number,
+    h: number
+  ) => {
+    return (spriteMap[name] = createSprite(image, x, y, w, h));
+  };
+
+  const addRotatedSprite = (
+    sprite: HTMLCanvasElement,
+    baseSpriteName: string,
+    n: number
+  ) => {
+    let rotated: HTMLCanvasElement = sprite;
+    for (let i = 0; i < n; i++) {
+      rotated = createRotatedImg(rotated);
+    }
+    addSprite(
+      `${baseSpriteName}_r${n}`,
+      rotated,
+      0,
+      0,
+      spriteWidth,
+      spriteHeight
+    );
+  };
+
+  const numColumns = image.width / spriteWidth;
+  const numRows = image.height / spriteHeight;
+
+  for (let i = 0; i < numRows; i++) {
+    for (let j = 0; j < numColumns; j++) {
+      // create original sprite: <baseSpriteName>
+      const baseSpriteName = `${spritePrefix}_${i * numColumns + j}`;
+      const sprite = addSprite(
+        baseSpriteName,
+        image,
+        j * spriteWidth,
+        i * spriteHeight,
+        spriteWidth,
+        spriteHeight
+      );
+
+      // create rotated sprites:<baseSpriteName>_rN
+      addRotatedSprite(spriteToCanvas(sprite), baseSpriteName, 1);
+      addRotatedSprite(spriteToCanvas(sprite), baseSpriteName, 2);
+      addRotatedSprite(spriteToCanvas(sprite), baseSpriteName, 3);
+
+      // create flipped sprite: <baseSpriteName>_f
+      addSprite(
+        `${baseSpriteName}${G_SPRITE_MOD_FLIPPED}`,
+        createFlippedImg(spriteToCanvas(sprite)),
+        0,
+        0,
+        spriteWidth,
+        spriteHeight
+      );
+    }
+  }
+};
+
+const loadImage = (imagePath: string): Promise<HTMLImageElement> => {
+  return new Promise(resolve => {
+    const img: HTMLImageElement = new Image();
+    img.onload = () => {
+      resolve(img);
+    };
+    img.src = imagePath;
```

Comment:
A little surprised by the ordering here. The promise resolves onload, and the src starts the load? On second thought, this makes much more sense.